### PR TITLE
Add new jax backend that calls the logp function through python

### PR DESCRIPTION
**Pull request by aseyboldt**
Getting jax support in nutpie without python overhead seems to be tricky, but this PR implements this by calling the jax logp functions through python.
This leads to relatively bad performance for very small functions, especially if many chains are running simultaneously (due to the python GIL). But for larger models this doesn't add much to the total runtime, so I think this is still quite useful.

The new backend can be used as this:
```python
compiled = nutpie.compile_pymc_model(model, backend="jax")
tr = nutpie.sample(compiled)
```

### CODE REVIEW

File: python/nutpie/compile_pymc.py [hunk 1]
```diff
@@ -258,22 +258,145 @@ def compile_pymc_model(model: "pm.Model", **kwargs) -> CompiledPyMCModel:
...
-        shape_info=shape_info,
-        logp_func=logp_fn_pt,
-        expand_func=expand_fn_pt,
+        dims=dims,
+        coords=coords,
     )
 
 
+def compile_pymc_model(
+    model: "pm.Model", *, backend="numba", gradient_backend=None, **kwargs
```
Comment (ricardoV94) [inline python/nutpie/compile_pymc.py:359]: String arguments is a place where I really like to have literal type hints
```suggestion
    model: "pm.Model", *, backend: Literal["numba", "jax"] ="numba", gradient_backend=None, **kwargs
```

File: python/nutpie/compile_pymc.py [hunk 2]
```diff
@@ -258,22 +258,145 @@ def compile_pymc_model(model: "pm.Model", **kwargs) -> CompiledPyMCModel:
...
+
+        logp_fn = logp_fn_jax_grad
+
+    shared_data = {}
+    shared_vars = {}
+    seen = set()
+    for val in [*logp_fn_pt.get_shared(), *expand_fn_pt.get_shared()]:
+        if val.name in shared_data and val not in seen:
+            raise ValueError(f"Shared variables must have unique names: {val.name}")
+        shared_data[val.name] = jax.numpy.asarray(val.get_value().copy())
```
Comment (ricardoV94) [inline python/nutpie/compile_pymc.py:311]: `get_value` copies by default (can be disabled with `borrow=True`)

File: python/nutpie/compile_pymc.py [hunk 3]
```diff
@@ -258,22 +258,149 @@ def compile_pymc_model(model: "pm.Model", **kwargs) -> CompiledPyMCModel:
...
-        dims=model.named_vars_to_dims,
-        _coords=coords,
-        _shapes={name: tuple(shape) for name, _, shape in zip(*shape_info)},
-        compiled_logp_func=logp_numba,
-        compiled_expand_func=expand_numba,
+    dims = model.named_vars_to_dims
+    return dims, coords
+
+
+def _compile_pymc_model_jax(model, *, gradient_backend=None, **kwargs):
```
Comment (twiecki) [inline python/nutpie/compile_pymc.py:265]: why is this jax specific (except for using that gradient if is indeed jax), seems like it would work with the c-backend (or any other) just fine.

Comment (aseyboldt) [reply python/nutpie/compile_pymc.py:265]: Yes, with the pytensor gradient backend it should already work with the c backend or the pytorch backends.
I'm not sure if it is a good idea to introduce the c backend as an option here, if we want to deprecate that soon anyway?

As for pytorch, I think we can add that in a separate PR (also needs changes to the optional dependencies to do properly). We can also add an explicit pytorch (or onnyx) backend in the rust code though, which shouldn't have those issues with the GIL.

Comment (aseyboldt) [reply python/nutpie/compile_pymc.py:265]: pytensor has no options to use the backend's autodiff machinery instead of its own, so if we want to do that, we need to bypass pytensor and access the generated function. In the PR we also bypass the shared-variable machinery of pytensor, so we access the compiled jax functions directly even if we use pytensor gradients.

This doesn't necessarily have to be like this, but I think I still prefer doing it that way, because the `with_data` machinery won't work otherwise, and we add more python overhead if we go through the pytensor function.
I'd also rather specify the optional dependencies correctly, and make sure things are tested, which is easier if we do this on a backend-by-backend basis.

Comment (twiecki) [reply python/nutpie/compile_pymc.py:265]: OK, I see, you're not calling the PyTensor function but the underlying JAX directly (through Python). That all makes sense then.
